### DISCUSSION TITLE
[Improvements]: Scheduled tasks

### DISCUSSION THREAD

**Discussion by andrasbacsai** [coollabsio/coolify#3152] [Improvement Requests]
Scheduled tasks added [here](https://github.com/coollabsio/coolify/pull/1607/), but there are a few improvements that will be added:

- [ ] Add email notifications for success/fail
- [ ] Trim executions to last ~20 or so to limit db growth for historic runs
- [ ] Execution time limits
- [ ] Display next run time, total execution time
- [ ] Probably some standardisation/consistency issues in here

**Comment by gausie** [comment]
I'd also love to be able to control the Scheduled Tasks as IAC by having a .crontab or similar file in the project's codebase. Is this a useful feature to track here or should it be considered separately?

**Reply by bioshazard** [reply]
I very much want this too. For now I am abusing env vars to make an easy copy-paste command to set it up in WebUI. IAC across coolify features would make it unstoppable

**Comment by peaklabs-dev** [comment]
- [ ] Also consider adding Slack notifications for failed tasks.
- [ ] Add tasks globally to coolify so for the coolify instance there can be cron jobs and for each container individually
- [ ] The pop-up input is too small for long cron task, consider using a wider area on the dashboard.
- [ ] Add a Timzone setting so we can select the timezone for each cron -> so it runs when we expect it

**Comment by gausie** [comment]
<img width="459" alt="Screenshot 2024-01-22 at 09 07 22" src="https://github.com/coollabsio/coolify/assets/179758/6f7305fe-0972-4ca2-89f0-77a1a4c9b9ab">

A task of mine seems to have got stuck a few days ago and there's no way to cancel it, nor did the task re-attempt the next day. Could we supply a timeout or something?

**Comment by aspantel** [comment]
We don't see the output of our scheduled tasks in our log drains. Is this not available?
Any ETA for this? 

We use scheduled tasks as health checks and getting their output is essential!

**Reply by the-leonat** [reply]
any news on this?

**Comment by jonnyscholes** [comment]
My "Recent executions" seem to change order on page load. This sometimes means the most recently executed task (even if it's currently running) don't show up in the list. Does this sound right?

Would you be open to a PR that ensures the recent executions are the 20 most recent ones?

**Comment by martmalo** [comment]
Also want to mention that failed cron jobs shows logs once I click on them, but successful ones does not. is this meant to be like that?

Edit:
Did not notice that someone mentioned this already.

**Comment by benjamin-steurer** [comment]
Is there any plan to integrate this with the log drains?

**Reply by the-leonat** [reply]
any news on this?

**Comment by HashimTheArab** [comment]
Is it possible to cancel stuck tasks?

**Reply by AlejandroAkbal** [reply]
Yes please, and set a maximum run time

**Comment by aspantel** [comment]
@andrasbacsai where does the container's FluentBit collect logs from? I am thinking maybe we can have a custom logback config to send our scheduled task logs to the container's FluentBit somehow. Can you share the default FluebtBit config when drians are configured? Thanks

**Reply by aspantel** [reply]
@andrasbacsai we have long running tasks and Coolify does not let them finish.
Is there some healthcheck mechanism for these tasks that our tasks are getting terminated by Coolify?

**Comment by da411d** [comment]
Is there way to set timezone for cron?

**Comment by pencilcheck** [comment]
Running long running tasks sometimes will not show output at all, and the status wouldn't reflect the process being called.

**Reply by wesselhuising** [reply]
Did you find any solution to this?

**Comment by pencilcheck** [comment]
I also noticed that scheduled tasks will run multiple instances of the same command at the same time in parallel which is causing a lot of problem as some command I want to only run once.

Is there a way to disable parallel running of the command? it is very annoying to find out the hard way as it slow the VPS to a crawl

**Comment by Rustypredator** [comment]
Also Essential for some applications would be to specify the user that executes the cron inside the container.  
For example Many PHP Applications such as Nextcloud etc. will need their cronjobs to be run by "www-data".  
I think this should be fairly easy by adding "-u {user}" on the docker exec command.

**Reply by Rustypredator** [reply]
Its not really as integrated as i would wish but this works (at least for nextcloud):
 - `su www-data -s /bin/sh -c "php cron.php"`  

It is necessary here to supply `-s /bin/sh` since www-data by default does not have a login-shell. the -s parameter tells the session which shell to use.
Hope this helps someone :wink:

**Reply by da411d** [reply]
Few weeks ago I had bug, when log file was created from cron context instead of http-server context, and then it was not writable for http-server context.

**Reply by sneko** [reply]
@peaklabs-dev can this be added to the TODO you listed above?

**Comment by moshetanzer** [comment]
Hey @andrasbacsai think it would be great to have a cron secret field on add takes that you can set a env car that will be passed as header when calling task to authenticate.

**Comment by dekaSW** [comment]
Hey @andrasbacsai 
I'd like to see support for step values like triggering a task at [every 15th minute](https://crontab.guru/#*/15_*_*_*_*) like so:

`*/15 * * * *`
Would this be easily doable?

**Comment by WFrancois** [comment]
Would love to have execution time limits!
Currently, if a task never stops because of an issue, new tasks will keep spawning until the host is unavailable.

**Comment by yannickbuntsma** [comment]
Scheduled tasks are so useful! 🙏 
I have a task that I would like to be able to turn on/off. Currently I'm "disabling" it by e.g. wrapping `yarn run-main` in an `echo` (`echo "yarn run-main"`) so it doesn't execute the actual command. 

Maybe I'm overlooking the feature, but otherwise it would be a possibly very small and useful addition (for me at least) 😄

**Comment by mjy78** [comment]
The Coolify Dragonfly template doesn't seem to expose a "Scheduled Tasks" configuration in the UI. Is there a way to enable this? Would be great to facilitate performing tasks like triggering snapshot saves via the redis-cli.

**Comment by wesselhuising** [comment]
Are there any updates on the development side of scheduled tasks?

**Comment by douglasg14b** [comment]
>  Trim executions to last ~20 or so to limit db growth for historic runs

Please provide controls for this. Different limits are desirable under different use cases.

**Comment by douglasg14b** [comment]
Can we capture logs from scheduled tasks?
